lib/engine/step/g_1860/tracker.rb

```diff
@@ -0,0 +1,279 @@
+                " for the #{ability.terrain} tile built by #{company.name}"
+            end
+          end
+        end
+
+        def border_cost(tile, entity)
+          hex = tile.hex
+          types = []
+
+          total_cost = tile.borders.dup.sum do |border|
+            next 0 unless (cost = border.cost)
+
+            edge = border.edge
+            neighbor = hex.neighbors[edge]
+            next 0 if !hex.targeting?(neighbor) || !neighbor.targeting?(hex)
+
+            types << border.type
+            tile.borders.delete(border)
+            neighbor.tile.borders.map! { |nb| nb.edge == hex.invert(edge) ? nil : nb }.compact!
+
+            ability = entity.all_abilities.find do |a|
+              (a.type == :tile_discount) && (border.type == a.terrain)
+            end
+            discount = ability&.discount || 0
+
+            if discount.positive?
+              @log << "#{entity.name} receives a discount of "\
+                "#{@game.format_currency(discount)} from "\
+                "#{ability.owner.name}"
+            end
+
+            cost - discount
+          end
+          [total_cost, types]
+        end
+
+        def check_track_restrictions!(entity, old_tile, new_tile, old_revenues)
+          return if @game.loading || !entity.operator?
+
+          changed_city = false
+          if old_tile.color != :white
+            # add requirement that paths/nodes be reachable with train
+            @game.game_error('Tile must be reachable with train') unless reachable_hex?(entity, new_tile.hex)
+            new_revenues = new_tile.nodes.select { |n| reachable_node?(entity, n) }.map(&:max_revenue).sort
+            changed_city = old_revenues != new_revenues
+          end
+
+          old_paths = old_tile.paths
+          used_new_track = old_paths.empty?
+
+          new_tile.paths.each do |np|
+            next unless @game.graph.connected_paths(entity)[np]
+            next if old_tile.color != :white && !reachable_path?(entity, np)
+
+            op = old_paths.find { |path| np <= path }
+            used_new_track = true unless op
+
+            next unless old_tile.color == :white
+
+            old_revenues = op&.nodes && op.nodes.map(&:max_revenue).sort
+            new_revenues = np&.nodes && np.nodes.map(&:max_revenue).sort
+            changed_city = true unless old_revenues == new_revenues
+          end
+
+          case @game.class::TRACK_RESTRICTION
+          when :permissive
+            true
+          when :city_permissive
+            @game.game_error('Must be city tile or use new track') if new_tile.cities.none? && !used_new_track
+          when :restrictive
+            @game.game_error('Must use new track') unless used_new_track
+          when :semi_restrictive
+            @game.game_error('Must use new track or change city value') if !used_new_track && !changed_city
+          else
+            raise
+          end
+        end
+
+        def potential_tiles(_entity, hex)
```

Comment:
isn't this just a copy of step/tracker.rb? did you just copy and paste it?

Reply:
Many of the methods are identical, but several have 1860-specific changes (distances, for example). I would have sub-classed Tracker, but it isn't a class. Can you override methods outside of a class?

Reply:
Hmmm. Tracker is a module included by the step/Track class. I needed my own Track because I needed to modify some of the methods in Tracker (i.e. I need an 1860-specific Tracker).  If I sub-class my 1860 Track from the generic one, any method redefined in my Tracker will be used by the other non-modified methods in the generic Track/Tracker, right?

I should be able to simply my code then.

Reply:
OK, just figured out that you *can* override methods outside of a class. The 1860 Tracker will now include the base Tracker